File: day11/solve.ts
import { sum, uniq } from "lodash";
import { solve } from "../runner/typescript";

function parser(input: string) {
  return input.split(" ").map(Number);
}

function blink(rock: number): number[] {
  if (rock === 0) return [1];

  const numDigits = Math.floor(Math.log10(rock)) + 1;
  if (numDigits % 2 === 0) {
    const halfDigits = numDigits / 2;
    const divisor = 10 ** halfDigits;
    const rightHalf = rock % divisor;
    const leftHalf = Math.floor(rock / divisor);
    return [leftHalf, rightHalf];
  }

  return [rock * 2024];
}

function part1(rocks: number[]): number {
  for (let i = 0; i < 25; i++) {
    rocks = rocks.flatMap(blink);
  }
  return rocks.length;
}

function part2(rocks: number[]): number {
  let counts = new Map(rocks.map((rock) => [rock, 1]));

  for (let i = 0; i < 75; i++) {
    const newCounts = new Map();
    counts.forEach((count, rock) =>
      blink(rock).forEach((newRock) =>
        newCounts.set(newRock, (newCounts.get(newRock) || 0) + count)
      )
    );
    counts = newCounts;
  }

  return sum([...counts.values()]);
}

solve({
  parser,
  part1,
  part2,
  part1Tests: [["125 17", 55312]],
});

File: day11/solve.py
from runner.python import solve

import math
from collections import defaultdict


def blink(rock: int) -> list[int]:
  if rock == 0:
    return [1]

  num_digits = math.floor(math.log10(rock)) + 1
  if num_digits % 2 == 0:
    half_digits = num_digits // 2
    divisor = 10**half_digits
    right_half = rock % divisor
    left_half = rock // divisor
    return [left_half, right_half]

  return [rock * 2024]


def part1(input: str) -> int:
  rocks = list(map(int, input.split()))
  for _ in range(25):
    rocks = [r for rock in rocks for r in blink(rock)]
  return len(rocks)


def part2(input: str) -> int:
  rocks = list(map(int, input.split()))
  counts = {rock: 1 for rock in rocks}

  for _ in range(75):
    new_counts = defaultdict(int)
    for rock, count in counts.items():
      for new_rock in blink(rock):
        new_counts[new_rock] += count
    counts = new_counts

  return sum(counts.values())


solve(
  part1,
  "125 17",
  55312,
  part2,
  "125 17",
  65601038650482,
)
